sdk/monitor/Azure.Monitor.Ingestion/src/LogsIngestionClient.cs

```diff
@@ -360,6 +375,56 @@ public virtual async Task<Response> UploadAsync<T>(string ruleId, string streamN
             return runningTasks.Select(_ => _.CurrentTask).Last().Result; //204 - response of last batch with header
         }
 
+        /// <summary>
+        /// test
+        /// </summary>
+        /// <param name="logsCount"></param>
+        /// <param name="options"></param>
+        /// <param name="response"></param>
+        /// <param name="cancellationToken"></param>
+        protected internal virtual async void OnException(int logsCount, UploadLogsOptions options, Response response, CancellationToken cancellationToken = default)
+        {
+            try
+            {
+                if (options == null)
+                {
+                    throw new RequestFailedException(response);
+                }
+                else
+                {
+                    await options.InvokeEvent(isRunningSynchronously: true, logsCount, response, cancellationToken).ConfigureAwait(false);
+                }
+            }
+            catch (Exception)
+            {
+            }
+        }
+
+        /// <summary>
+        /// test
+        /// </summary>
+        /// <param name="logsCount"></param>
+        /// <param name="options"></param>
+        /// <param name="response"></param>
+        /// <param name="cancellationToken"></param>
+        protected internal virtual async Task OnExceptionAsync(int logsCount, UploadLogsOptions options, Response response, CancellationToken cancellationToken = default)
```

Comment:
Instead of taking the response and logs count, should this take the UploadFailedArgs?

Reply:
I made the Args in the InvokeEvent method on the UploadLogsOptions class to follow the sample: https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/tests/samples/FakeClients/AlarmClient.cs#L35. The Args use the isRunningSynchronously param